proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,252 @@
+
+## `bool.lower_iMM`
+
+This instruction has type `[bool] -> iMM`.
+
+The output is `0b1` zero extended to the desired bitwidth iff the input is true.
+
+## `sNN.lift_iMM`
+
+This instruction has type `iMM` -> `sNN`.
+
+The output is the value of `iMM` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `sNN.lower_iMM`
+
+This instruction has type `[sNN] -> [iMM]`
+
+The output is the value of `sNN` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lift_iMM`
+
+This instruction has type `iMM` -> `uNN`.
+
+The output is the value of `iMM` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lower_iMM`
+
+This instruction has type `[uNN] -> [iMM]`
+
+The output is the value of `uNN` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `string.lift_memory`
+
+The full syntax of this instruction is:
+```
+string.lift_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, $len: i32] -> [string]
+```
+
+The output is a sequence of unicode code points taken by interpreting `store.memory[$memidx].bytes[$base .. $base + $len]` with `#encoding`.
+
+This instructions traps if the range `$base .. $base + $len` is out-of-bounds, or the byte sequence is not valid under `#encoding`.
+
+The syntax for supported encodings is:
+```
+#encoding ::=
+  | utf8
+```
+
+Currently, only UTF8 is supported as an encoding. UTF16 may be added in the future.
+
+## `string.lower_memory`
+
+The full syntax of this instruction is:
+```
+string.lower_memory #memidx #encoding $alloc: #funcidx
+```
+
+The type of this instruction is:
+```
+[string] -> [i32, i32]
+
+Where:
+  * $alloc has a valid functype of [i32] -> [i32]
+```
+
+The instruction will execute the following steps:
+```
+  1. Let $length be the bytes required to encode the string with #encoding
+  2. Call $alloc with $length and receive an offset into #memidx
+  3. Encode the string into the offset at memory
+```
+
+## `record.lift`
+
+The full syntax of this instruction is:
+```
+record.lift #typeidx
+```
+
+The type of this instruction is:
+```
+[$a, $b, ..] -> [$type]
+
+Where:
+  * $type is a type of the form (record $fields) taken from #typeidx
+  * $a, $b, .. are the #interface-valtype's for each field in $fields
+```
+
+The output is a record interface value with fields given by the inputs.
+
+## `record.lower`
+
+The full syntax of this instruction is:
+```
+record.lower #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [$fields]
+
+Where:
+  * $type is a valid type of the form (record $fields) taken from #typeidx
+```
+
+The outputs are the fields of the input `record`.
+
+## `array.lift_memory`
+
+This is a structured instruction with the following syntax:
```

Comment:
use a name that reflects what is going on: iterating through the elements of the array.
also need 
array.count
and
array.size
so that memory can be managed properly.

Reply:
What are `array.count/array.size` needed for?

Reply:
see below

Reply:
`array.count` and `string.size` were added.